test/e2e/e2e_test.go

```diff
@@ -0,0 +1,179 @@
+			Name: serviceName,
+			Annotations: map[string]string{
+				servingcert.ServingCertSecretAnnotation: secretName,
+			},
+		},
+		Spec: v1.ServiceSpec{
+			Ports: []v1.ServicePort{
+				{
+					Name: "tests",
+					Port: 8443,
+				},
+			},
+		},
+	})
+	if err != nil {
+		return "", "", err
+	}
+	return serviceName, secretName, nil
+}
+
+func pollForServiceServingSecret(client *kubernetes.Clientset, secretName, namespace string) error {
+	return wait.PollImmediate(time.Second, 10*time.Second, func() (bool, error) {
+		_, err := client.CoreV1().Secrets(namespace).Get(secretName, metav1.GetOptions{})
+		if err != nil && errors.IsNotFound(err) {
+			return false, nil
+		}
+		if err != nil {
+			return false, err
+		}
+		return true, nil
+	})
+}
+
+func cleanupServiceSignerTestObjects(client *kubernetes.Clientset, secretName, serviceName, namespace string) {
+	client.CoreV1().Secrets(namespace).Delete(secretName, &metav1.DeleteOptions{})
+	client.CoreV1().Services(namespace).Delete(serviceName, &metav1.DeleteOptions{})
+	client.CoreV1().Namespaces().Delete(namespace, &metav1.DeleteOptions{})
+}
+
+func TestE2E(t *testing.T) {
+	// use /tmp/admin.conf (placed by ci-operator) or KUBECONFIG env
+	confPath := "/tmp/admin.conf"
+	if conf := os.Getenv("KUBECONFIG"); conf != "" {
+		confPath = conf
+	}
+
+	// load client
+	client, err := clientcmd.LoadFromFile(confPath)
+	if err != nil {
+		t.Fatalf("error loading config: %v", err)
+	}
+	adminConfig, err := clientcmd.NewDefaultClientConfig(*client, &clientcmd.ConfigOverrides{}).ClientConfig()
+	if err != nil {
+		t.Fatalf("error loading admin config: %v", err)
+	}
+	adminClient, err := kubernetes.NewForConfig(adminConfig)
+	if err != nil {
+		t.Fatalf("error getting admin client: %v", err)
+	}
+
+	// the service-ca-operator and controllers should be running as a stock OpenShift component. our first test is to
+	// verify that all of the components are running.
+	if !hasPodWithPrefixName(adminClient, serviceCAOperatorPodPrefix, serviceCAOperatorNamespace) {
+		t.Fatalf("%s not running in %s namespace", serviceCAOperatorPodPrefix, serviceCAOperatorNamespace)
+	}
+	if !hasPodWithPrefixName(adminClient, apiInjectorPodPrefix, serviceCAControllerNamespace) {
+		t.Fatalf("%s not running in %s namespace", apiInjectorPodPrefix, serviceCAControllerNamespace)
+	}
+	if !hasPodWithPrefixName(adminClient, configMapInjectorPodPrefix, serviceCAControllerNamespace) {
+		t.Fatalf("%s not running in %s namespace", configMapInjectorPodPrefix, serviceCAControllerNamespace)
+	}
+	if !hasPodWithPrefixName(adminClient, caControllerPodPrefix, serviceCAControllerNamespace) {
+		t.Fatalf("%s not running in %s namespace", caControllerPodPrefix, serviceCAControllerNamespace)
+	}
+
+	// test the main feature. annotate service -> created secret
+	t.Run("serving-cert-annotation", func(t *testing.T) {
+		ns := createTestNamespace(adminClient)
+		if len(ns) == 0 {
+			t.Fatalf("could not create test namespace")
+		}
```

Comment:
Should we clean up the namespace as well?

Reply:
we do in cleanupServiceSignerTestObjects()

Reply:
What about in the situation that `createServinCertAnnotatedService` fails below?

Reply:
Good catch, fixed that.